File: dp/leetcode42.java
package dp;

import java.util.Scanner;

public class leetcode42 {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] height = new int[n];
        for (int i = 0; i < n; i++) {
            height[i] = in.nextInt();
        }
        System.out.println(trap(height));
        in.close();
    }

    static int trap(int[] height) {
        if (height.length == 0)
            return 0;
        int res = 0;
        int left = 0;
        int right = height.length - 1;
        int lmax = height[left];
        int rmax = height[right];
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] <= lmax)
                    res = res + lmax - height[left];
                else
                    lmax = height[left];
                left++;
            } else {
                if (height[right] <= rmax)
                    res = res + rmax - height[right];
                else
                    rmax = height[right];
                right--;
            }
        }
        return res;
    }
}

File: common_list/leetcode42.go
package common_list

// 给定 n 个非负整数表示每个宽度为 1 的柱子的高度图，计算按此排列的柱子，下雨之后能接多少雨水。

// 思路：双指针法
// 记录左右最高值，左值比右值小时，正序遍历，加上空余面积，右值比左值小时，逆序遍历
func Trap(height []int) int {
	res := 0
	lmax, rmax := 0, 0
	for l, r := 0, len(height)-1; l <= r; {
		if height[l] < height[r] {
			if height[l] > lmax {
				lmax = height[l]
			} else {
				res += lmax - height[l]
			}
			l++
		} else {
			if height[r] > rmax {
				rmax = height[r]
			} else {
				res += rmax - height[r]
			}
			r--
		}
	}
	return res
}
